File: labs/week-06/assignment02.scala
package main.scala.chapter3

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._

object assignment02 {
    def main(args : Array[String]){
        val spark = (SparkSession
                .builder
                .appName("assignment02")
                .getOrCreate())
 

    val fireSchema = StructType(Array(StructField("device_id", IntegerType, false),
    StructField("device_name", StringType, false),
    StructField("ip", StringType, false),
    StructField("cca2", StringType, false),
    StructField("cca3", StringType, false), 
    StructField("cn", StringType, false),
    StructField("latitude", StringType, false),
    StructField("longitude", DoubleType, false),
    StructField("scale", StringType, false),
    StructField("temp", IntegerType, false), 
    StructField("humidity", IntegerType, false), 
    StructField("battery_level", IntegerType, false), 
    StructField("c02_level", IntegerType, false), 
    StructField("lcd", StringType, false), 
    StructField("timestamp", LongType, false)))    

    val File_Path="/home/vagrant/learningSparkV2/databricks-datasets/learning-spark-v2/iot-devices/iot_devices.json"

    val DataF = spark.read.schema(fireSchema)
        .json(File_Path)

    DataF.show(false)
    println(DataF.printSchema)

    DataF.createOrReplaceTempView("framedf")

    //Detect failing devices with battery levels below a threshold.
    val First = spark.sql("Select device_name,device_id,battery_level from framedf where battery_level < 10")
    First.show(20, false)

    //Identify offending countries with high levels of C02 emissions.
    val Second = spark.sql("Select cca3,c02_level from framedf where c02_level = 1599")
    Second.show(20, false)

    //Compute the min and max values for temperature, battery level, C02, and humidity
    val Third = spark.sql("Select min(temp),min(battery_level),min(c02_level),min(humidity),max(temp),max(battery_level),max(c02_level),max(humidity) from framedf")
    Third.show(20, false)

    //Sort and group by average temperature, Co2, humidity, and country
    val Four = spark.sql("Select AVG(temp),AVG(c02_level),AVG (humidity),cca3 from framedf group by cca3")
    Four.show(20, false)
    
    }
}

File: labs/week-06/assignment-02.py

from pyspark.sql import SparkSession 
from pyspark.sql.types import *
from pyspark.sql.functions import col
from pyspark.sql.functions import max
from pyspark.sql.functions import to_timestamp
from pyspark.sql.functions import year
from pyspark.sql.functions import month
from pyspark.sql.functions import weekofyear
from pyspark.sql.functions import desc

spark = (SparkSession
.builder
.appName("assignment-02")
.getOrCreate())


fireSchema = StructType([StructField('CallNumber', IntegerType(), True),
 StructField('UnitID', StringType(), True),
 StructField('IncidentNumber', IntegerType(), True),
 StructField('CallType', StringType(), True), 
 StructField('CallDate', StringType(), True), 
 StructField('WatchDate', StringType(), True),
 StructField('CallFinalDisposition', StringType(), True),
 StructField('AvailableDtTm', StringType(), True),
 StructField('Address', StringType(), True), 
 StructField('City', StringType(), True), 
 StructField('Zipcode', IntegerType(), True), 
 StructField('Battalion', StringType(), True), 
 StructField('StationArea', StringType(), True), 
 StructField('Box', StringType(), True), 
 StructField('OriginalPriority', StringType(), True), 
 StructField('Priority', StringType(), True), 
 StructField('FinalPriority', IntegerType(), True), 
 StructField('ALSUnit', BooleanType(), True), 
 StructField('CallTypeGroup', StringType(), True),
 StructField('NumAlarms', IntegerType(), True),
 StructField('UnitType', StringType(), True),
 StructField('UnitSequenceInCallDispatch', IntegerType(), True),
 StructField('FirePreventionDistrict', StringType(), True),
 StructField('SupervisorDistrict', StringType(), True),
 StructField('Neighborhood', StringType(), True),
 StructField('Location', StringType(), True),
 StructField('RowID', StringType(), True),
 StructField('Delay', FloatType(), True)])

File="/home/vagrant/learningSparkV2/databricks-datasets/learning-spark-v2/sf-fire/sf-fire-calls.csv"

fireDF = spark.read.schema(fireSchema).option("header", "true").csv(File)
print(fireDF.show())
print(fireDF.printSchema())

result = fireDF.withColumn("IncidentDate", to_timestamp(col("CallDate"), "MM/dd/yyyy")).drop("CallDate")
result.select(year("IncidentDate"))

 # What were all the different types of fire calls in 2018?
(result
 .select("CallType")
 .where((year("IncidentDate"))=="2018")
 .groupby("CallType")
 .count()
 .show(truncate = False))

 # What months within the year 2018 saw the highest number of fire calls?
(result
 .select("IncidentDate")
 .where((year("IncidentDate"))=="2018")
 .groupby(month("IncidentDate"))
 .count()
 .orderBy('count',ascending=False)
 .show())

# Which neighborhood in San Francisco generated the most fire calls in 2018?
(result
.select("Neighborhood")
.where(((year("IncidentDate")) == "2018") & (col("City") == "San Francisco"))
.groupby(("Neighborhood"))
.count()
.show(30,truncate = False))

# Which neighborhoods had the worst response times to fire calls in 2018?
(result
.select("Neighborhood", "Delay")
.where(((year("IncidentDate")) == "2018") & (col("City") == "San Francisco"))
.sort(col("Delay").desc())
.show(5,truncate=False))

# Which week in the year in 2018 had the most fire calls?
(result
 .select("IncidentDate")
 .where((year("IncidentDate"))=="2018")
 .groupby(weekofyear("IncidentDate"))
 .count()
 .show(40))

# How can we use Parquet files or SQL tables to store this data and read it back?
#parquet_path = "./"
parquet_path = "~"
result.write.format("parquet").save(parquet_path)

# Is there a correlation between neighborhood, zip code, and number of fire calls?
(fireDF
 .select("Neighborhood","Zipcode")
 .stat.corr("Neighborhood","Zipcode")
 .count()
 .orderBy('count')
 .show())




















